### PR TITLE
feat: Added stft Function Spectral Ops to PyTorch Frontend Close #21305

### LINKED ISSUE #21305: stft

### PR DESCRIPTION
**Pull request by Juligee**
Close #21305
#Add Spectral Ops to PyTorch Frontend #15

from typing import (
    List, Tuple, Optional, Union, Any, Sequence, TYPE_CHECKING
)

import torch
from torch._C import _add_docstr
import torch.backends.opt_einsum as opt_einsum
import torch.nn.functional as F
from ._lowrank import svd_lowrank, pca_lowrank
from .overrides import (
    has_torch_function, has_torch_function_unary, has_torch_function_variadic,
    handle_torch_function)
from ._jit_internal import boolean_dispatch
from ._jit_internal import _overload as overload

Tensor = torch.Tensor
from torch import _VF

__all__ = [
    'atleast_1d',
    'atleast_2d',
    'atleast_3d',
    'align_tensors',
    'broadcast_shapes',
    'broadcast_tensors',
    'cartesian_prod',
    'block_diag',
    'cdist',
    'chain_matmul',
    'einsum',
    'istft',
    'lu',
    'norm',
    'meshgrid',
    'pca_lowrank',
    'split',
    'stft',
    'svd_lowrank',
    'tensordot',
    'unique',
    'unique_consecutive',
]

#stft computes Fourier transform of short overlapping windows of the input. 
def stft(input: Tensor, n_fft: int, hop_length: Optional[int] = None,
         win_length: Optional[int] = None, window: Optional[Tensor] = None,
         center: bool = True, pad_mode: str = 'reflect', normalized: bool = False,
         onesided: Optional[bool] = None,
         return_complex: Optional[bool] = None) -> Tensor:

### CODE REVIEW

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
Hi @Juligee 

You have added the functions in on the main repo, the main folder structure should be followed.
Kindly look into these two deep dives to fix this:
https://unify.ai/docs/ivy/ovhttps://unify.ai/docs/ivy/overview/deep_dive/ivy_frontends_tests.htmlerview/deep_dive/ivy_frontends.html

Also follow the structure of implementation we use in the documentation which I linked you to above. Use hypothesis for testing your code.

Thanks!

**Comment by Juligee** [comment]
Good morning, @zaeemansari70. 

Okay, I will make the corrections as requested.

Thank you

**Comment by Juligee** [comment]
Hi @zaeemansari70.
 The link you gave above is "404"
[https://unify.ai/docs/ivy/ovhttps://unify.ai/docs/ivy/overview/deep_dive/ivy_frontends_tests.htmlerview/deep_dive/ivy_frontends.htm](https://unify.ai/docs/ivy/ovhttps://unify.ai/docs/ivy/overview/deep_dive/ivy_frontends_tests.htmlerview/deep_dive/ivy_frontends.html)

**Comment by Juligee** [comment]
Dear @zaeemansari70, 

I have made update to my pull request as required. Kindly update with further reviewS.

Best Regards

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/functional/frontends/torch/spectra_ops.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+    if not ivy.is_complex_dtype(input):
+        raise ivy.exceptions.IvyError(
+            "view_as_real is only supported for complex tensors"
+        )
+    re_part = ivy.real(input)
+    im_part = ivy.imag(input)
+    return ivy.stack((re_part, im_part), axis=-1)
+
+
+@to_ivy_arrays_and_back
```
Comment (zaeemansari70) [inline ivy/functional/frontends/torch/spectra_ops.py:433]: This should be the only function you should implement, can you please un-commit all the other functions, they are all present in `miscellaneous_ops.py`thanks!

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py [hunk 2]
```diff
@@ -0,0 +1,1620 @@
...
+        frontend=frontend,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        input=np.asarray(x[0], dtype=input_dtype[0]),
+    )
+
+
+# stft
+@handle_frontend_test(
```
Comment (zaeemansari70) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py:1590]: Same for this, we only need this test.

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
Kindly add just the implementation which I highlighted to `ivy/functional/frontends/torch/spectral_ops.py` and the test in the respective test file.

File: ivy/functional/frontends/torch/spectra_ops.py (hunk 1 above)
Comment (Juligee) [reply ivy/functional/frontends/torch/spectra_ops.py:433]: Okay, I will make the changes as directed.
Thank you very much for the update.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py (hunk 2 above)
Comment (Juligee) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py:103]: Good day, @zaeemansari70 ,

I have made the requested changes. Thank you.
However, I have further questions;
1. >>
ModuleNotFoundError                       Traceback (most recent call last)
[<ipython-input-6-6b3ffb48f7e4>](https://localhost:8080/#) in <cell line: 5>()
      3 
      4 import numpy as np
----> 5 from hypothesis import assume, strategies as st
      6 import hypothesis.extra.numpy as nph
      7 

ModuleNotFoundError: No module named 'hypothesis'

---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.
---------------------------------------------------------------------------
 How can I find the issue it has?

2.>> How can I find a compiled list of ivy function types and their tests codes. The document has some few examples like ivy.tan( ), ivy.tan( )

Thank you.

Comment (zaeemansari70) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py:103]: You have not properly configured the environment, I'd suggest follow this according to your use case https://unify.ai/docs/ivy/overview/contributing/setting_up.html Thanks!

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    # Output size is an optional parameter accepted by Torch for optimisation
+    use_output_size = draw(st.booleans())
+    output_size = np.sum(repeats) if use_output_size else None
+
+    return [values_dtype, repeats_dtype], values, repeats, axis, output_size
+
+
+# atleast_1d
+@handle_frontend_test(
```
Comment (zaeemansari70) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py:63]: This should be uncommited, the only changes that show should be for the `stft` function

**Review by zaeemansari70 — CHANGES_REQUESTED** [review]
Let me know if you have any questions, Thanks! :slightly_smiling_face:

File: ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py (hunk 3 above)
Comment (Juligee) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_spectra_ops.py:63]: Good, evening, okay. It's really getting clearer with your help! 
Thank you.

**Comment by Juligee** [comment]
Dear @zaeemansari70, 

I having issue setting up my environment - first, I tried setting up for free as it is in the IVY doc. -   using Vscode and Docker and I decided to avoid the packages installations and used the Github codespace. 
I get this error:

"This codespace is currently running in recovery mode due to a configuration error. 
Please review the creation logs, 
update your dev container configuration as needed, 
and run the "Rebuild Container" command to rectify."

Could you please help assist with this error?

**Comment by Juligee** [comment]
Good evening @zaeemansari70,

I was able to resolve I this error:

"This codespace is currently running in recovery mode due to a configuration error.
Please review the creation logs,
update your dev container configuration as needed,
and run the "Rebuild Container" command to rectify."

I had to delete the previous codespace and create a new one and had follow the Github configuration and setup for Vscode.
My ivy_frontend and ivy_test codes for stft function are also now in the correct file explorer locations as directed.

Thank you.

**Comment by Dharshannan** [comment]
Hi @Juligee, I have recently added the stft function to the ivy Functional API, you can update this PR to call the ivy.stft function directly, are you still working on this PR btw?

**Comment by Juligee** [comment]
Good morning,

@Dharshannan, yes, I am still working on this frontend ivy API. I was working to get the right ivy function to test with the native backend. Also, I am learning about how to set up environment for testing.

Thank you very much for the updates.

**Comment by Juligee** [comment]
> ivy.stft

@Dharshannan, good evening. I checked the main file for frontend - spectra_ops.py, but it showed 
image

Kindly send me the url to find the ivy.stft function. 

Thank you

**Comment by Dharshannan** [comment]
@Juligee , the stft function is added to the functional API, so meaning it is in all the backends including Torch, the array, container instances and the functional/layers.py script. For YOUR CASE you can directly call the ivy.stft within the torch frontend function as such:

@to_ivy_arrays_and_back
def stft(
    signals,
    frame_length,
    frame_step,
    fft_length=None,
    window_fn=None,
    pad_end=False,
    name=None,
):
    signals = ivy.asarray(signals)
    return ivy.stft(
        signals,
        frame_length,
        frame_step,
        fft_length=fft_length,
        window_fn=window_fn,
        pad_end=pad_end,
        name=name,
    )

I would recommend following this format (*as it was difficult to superset the behavior of the stft function across all backends due to how padding and the other arguments are handled, hence when I implemented the stft into the functional API I used the Tensorflow format as the ground truth to superset the other backend stft functions), as of now, if you need help with the test function let me know as well. For reference of the stft in the functional API here is the merged PR link:
https://github.com/unifyai/ivy/pull/22581

**Comment by Juligee** [comment]
> @Juligee , the stft function is added to the functional API, so meaning it is in all the backends including Torch, the array, container instances and the functional/layers.py script. For YOUR CASE you can directly call the ivy.stft within the torch frontend function as such:
> 
> @to_ivy_arrays_and_back def stft( signals, frame_length, frame_step, fft_length=None, window_fn=None, pad_end=False, name=None, ): signals = ivy.asarray(signals) return ivy.stft( signals, frame_length, frame_step, fft_length=fft_length, window_fn=window_fn, pad_end=pad_end, name=name, )
> 
> I would recommend following this format (*as it was difficult to superset the behavior of the stft function across all backends due to how padding and the other arguments are handled, hence when I implemented the stft into the functional API I used the Tensorflow format as the ground truth to superset the other backend stft functions), as of now, if you need help with the test function let me know as well. For reference of the stft in the functional API here is the merged PR link: #22581

Good day @Dharshannan, 

Thank you for this detailed explanation and guide. I would really move forward with the stft frontend test, and I am happy your assistance would be helpful. 

Thank you.

**Comment by Juligee** [comment]
Dear @Dharshannan 

Please you give an idea of how to clear this check error?
"Trigger lint format / PR Lint Bot / format-code (pull_request) Failing after 3s"

image

Thank you very much

**Comment by Dharshannan** [comment]
@Juligee , I am not too sure as lint check is a formatting error, do you have pre-commit installed? Probably best to ask someone else more experienced in lint errors, as I am only a volunteer.

**Comment by Juligee** [comment]
> @Juligee , I am not too sure as lint check is a formatting error, do you have pre-commit installed? Probably best to ask someone else more experienced in lint errors, as I am only a volunteer.

Good day @Dharshannan,

Okay, thank you for the feedback. I am really thankful for the assistance so far. 

Best regards.

**Comment by Daniel4078** [comment]
I found that all the relevent tests are currently failing for : E   ImportError: Error: The function 'stft' could not be found within the module 'ivy.functional.frontends.torch'.

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Pull request closed without merging**